### ISSUE TITLE
Possibility to start a defined program

### ISSUE THREAD

**Issue by Phoenix-DH** [humbertogontijo/homeassistant-roborock#419]
In the roborock app you can define programs, ways on how the robo should clean room by room:
![IMG_5999](https://github.com/humbertogontijo/homeassistant-roborock/assets/44058592/435cb87c-e60f-4e29-adc2-a19524ee0040)

Is there a way to implememt this in the integration?

Would be really great.

Thanks.

**Comment by olivergruber** [comment]
This is called "routines" in the english app.

I second this, it would be very convenient.

**Comment by humbertogontijo** [comment]
I’ve tried implementing it. But there are some information stored in the app itself. And the first thing the app does is erase any unknown routine. So anything done in HA would be erased once you open the app

**Comment by cs42** [comment]
> I’ve tried implementing it. But there are some information stored in the app itself. And the first thing the app does is erase any unknown routine. So anything done in HA would be erased once you open the app

I think it's not so important to create or edit routines. But it would really be convenient to start already defined routines. Then one could use HA to determine, when to run them. As you can have multiple routines an schedule them, I'd expect that it should be possible to start them sending a command to the robot (or the cloud?).

**Comment by Phoenix-DH** [comment]
Would it not be enought and the possibility for this scenario just to use the information out of the app/cloud? Means just to trigger some routine-ids? Like with changing mop intense …

**Comment by humbertogontijo** [comment]
I can check it out. But that routine would not be named accordingly since that information is not stored in the robot

**Comment by Phoenix-DH** [comment]
Many thanks!

**Comment by StaleLoafOfBread** [comment]
This is a duplicate of https://github.com/humbertogontijo/homeassistant-roborock/issues/169

**Comment by CH1PSET1** [comment]
Yes, the implementation of this function would be simply amazing. She is greatly missed.

**Comment by Phoenix-DH** [comment]
Are there some news? Thanks.

**Comment by edmooredop** [comment]
Hey there, just to give my support for this idea. The Routines section of the Roborock app allows quite a lot of control that's not available otherwise - for example, you can specify a zone to clean (I use this to define an area around where our cats eat which needs more frequent cleaning), and you can tell it to repeat the task several times if needed.

Would love to be able to trigger an existing routine via the HA integration - for example if I know via other means that the cat has eaten recently, start the roborock routine :)

At the moment the only option would be to split off the area as a seperate room in Roborock's map which creates a bit of a mess when you just want to clean everywhere all together.

**Comment by Lawris** [comment]
I second this.

But I think it would not be that stable, tho. Every time the map has a slight change in the app, the routines are greyed out and not executable anymore, this might affect HA as well.

**Comment by arne182** [comment]
I have something that works. Not perfect but here it is:

alias: Add Selected Room to Queue
description: ""
trigger:
  - platform: state
    entity_id:
      - input_button.add_to_queue
action:
  - variables:
      selected_room: "{{ states('input_select.room_to_clean').split(',') }}"
      room_name: "{{ selected_room[0] }}"
      segment_id: "{{ selected_room[1] }}"
      repeats: "{{ states('input_number.repeats') | int }}"
  - service: input_text.set_value
    target:
      entity_id: input_text.segment_queue
    data:
      value: >
        {% set queue = states('input_text.segment_queue') %} {% set new_entry =
        segment_id ~ ':' ~ repeats %} {{ queue + ',' + new_entry if queue | trim
        != '' else new_entry }}
  - service: automation.trigger
    target:
      entity_id: automation.check_and_clean_queue
    data:
      skip_condition: false
mode: single

alias: Check and Clean Queue
description: Checks the queue every minute and cleans if robot is docked.
trigger:
  - platform: time_pattern
    minutes: /1
condition:
  - condition: template
    value_template: "{{ states('input_text.segment_queue') | trim != '' }}"
  - condition: or
    conditions:
      - condition: state
        entity_id: vacuum.robot_city
        state: charging
        attribute: stateName
        for:
          hours: 0
          minutes: 0
          seconds: 0
      - condition: numeric_state
        entity_id: vacuum.robot_city
        attribute: cleanPercent
        above: 98
        enabled: false
      - condition: state
        entity_id: vacuum.robot_city
        state: charging
        attribute: stateName
        for:
          hours: 0
          minutes: 0
          seconds: 0
      - condition: state
        entity_id: vacuum.robot_city
        state: idle
        attribute: stateName
        for:
          hours: 0
          minutes: 0
          seconds: 0
      - condition: state
        entity_id: vacuum.robot_city
        state: paused
        attribute: stateName
        for:
          hours: 0
          minutes: 0
          seconds: 0
action:
  - variables:
      segment_info: "{{ states('input_text.segment_queue').split(',')[0] }}"
      segment_id: "{{ segment_info.split(':')[0] }}"
      repeats: "{{ segment_info.split(':')[1] }}"
  - service: roborock.vacuum_clean_segment
    target:
      entity_id: vacuum.robot_city
    data:
      repeats: "{{ repeats }}"
      segments: "{{ segment_id }}"
  - service: input_text.set_value
    target:
      entity_id: input_text.segment_queue
    data:
      value: "{{ states('input_text.segment_queue').split(',')[1:] | join(',') }}"
mode: single
![Screenshot_20240128_185422_Home Assistant](https://github.com/humbertogontijo/homeassistant-roborock/assets/844102/cd461708-1b2a-4f1e-90e4-0b77ffa0a64c)
The segment queue is then run one after another like the routines

**Comment by Phoenix-DH** [comment]
Great idea … but I think oversized for me.
Does anybody know if I can just set this value via command?

![IMG_7228](https://github.com/humbertogontijo/homeassistant-roborock/assets/44058592/00458589-9f04-4574-b8c5-bf0ffdb8b835)

Means just start the Imdividuell thing?
Because then I can realize it.
Just change the card and then set this.

**Comment by regevbr** [comment]
please add a thumbs up for https://github.com/Python-roborock/python-roborock/pull/317 to get this rolling
